docs/deployment_configuration.md
# Configuration via environment variables for deployment of Child Chain, Watcher and Watcher Info releases

***Child Chain, Watcher and Watcher Info***

- "PORT" - Child Chain or Watcher API port. Defaults to 9656 for Child Chain and 7434 for Watcher.
- "HOSTNAME" - server domain name of Child Chain or Watcher. *mandatory*
- "DD_DISABLED" - boolean that allows you to turn on or of Datadog metrics. Defaults to true.
- "APP_ENV" - environment name in which the the application was deployed. *mandatory*
- "DB_PATH" - directory of the KV db. *mandatory*
- "ETHEREUM_RPC_URL" - address of Geth or Parity instance. *mandatory*
- "ETH_NODE" - Geth, Parity or Infura. *mandatory*
- "SENTRY_DSN" - if not set, Sentry is disabled.
- "DD_HOSTNAME" - Datadog hostname.
- "DD_PORT" - Datadog agent UDP port for metrics.
- "DD_APM_PORT" - Datadog TCP port for APM.
- "BATCH_SIZE" - Datadog batch size for APM.
- "SYNC_THRESHOLD" - Datadog sync threshold for APM.
- "ETHEREUM_BLOCK_TIME_SECONDS" - Should mirror Ethereum network's setting, defaults to 15 seconds.
- "ETHEREUM_EVENTS_CHECK_INTERVAL_MS" - the frequency of HTTP requests towards the Ethereum clients and scanning for interested events. Should be less then average block time (10 to 20 seconds) on Ethereum mainnet.
- "ETHEREUM_STALLED_SYNC_THRESHOLD_MS" - the threshold before considering an unchanging Ethereum block height to be considered a stalled sync. Should be slightly larger than the expected block time.
- "LOGGER_BACKEND" - Ink or console. Ink will encode logs as json (useful for Datadog). Console will use the default elixir Logger backend. Default is Ink.

***Child Chain only***

- "BLOCK_SUBMIT_MAX_GAS_PRICE" - The maximum gas price to use for block submission. The first block submission after application boot will use the max price. The gas price gradually adjusts on subsequent blocks to reach the current optimum price . Defaults to `20000000000` (20 Gwei).
- "BLOCK_SUBMIT_EVERY_NTH" - submits block every `N` Ethereum blocks (unless block is empty). Default is 1 (plasma block in each Ethereum block).
- "BLOCK_SUBMIT_AFTER_N_TXS" - triggers block submission quicker than above n-Ethereum blocks if there are already `N` pending transactions.
- "FEE_ADAPTER" - The adapter to use to populate the fee specs. Either `file` or `feed` (case-insensitive). Defaults to `file` with an empty fee specs.
- "FEE_CLAIMER_ADDRESS" - 20-bytes HEX-encoded string of Ethereum address of Fee Claimer.
- "FEE_BUFFER_DURATION_MS" - Buffer period during which a fee is still valid after being updated.
- "FEE_SPECS_FILE_PATH" - The path to the fee specs file including the file name.  Only applicable when `FEE_ADAPTER=file`.
- "FEE_FEED_URL" - URL to fee feed service. Only applicable when `FEE_ADAPTER=feed`.
- "FEE_CHANGE_TOLERANCE_PERCENT" - Positive integer describes significance of price change. When price in new reading changes above tolerance level, prices are updated immediately. Otherwise update interval is preserved. Only applicable when `FEE_ADAPTER=feed`.
- "STORED_FEE_UPDATE_INTERVAL_MINUTES" - Positive integer describes time interval in minutes. The updates of token prices are carried out in update intervals as long as the changes are within tolerance. Only applicable when `FEE_ADAPTER=feed`.

***Watcher and Watcher Info only***

- "CHILD_CHAIN_URL" - Location of the Child Chain API. *mandatory*
- "EXIT_PROCESSOR_SLA_MARGIN" - Number of Ethereum blocks since start of an invalid exit, before `unchallenged_exit` is reported to prompt to mass exit. Must be smaller than "MIN_EXIT_PERIOD_SECONDS", unless "EXIT_PROCESSOR_SLA_MARGIN_FORCED=TRUE".

***Watcher Info only***

- "DATABASE_URL" - Postgres address *mandatory*
- "WATCHER_INFO_DB_POOL_SIZE" - The size of the database connection pool. Defaults to `10`.
- "WATCHER_INFO_DB_POOL_QUEUE_TARGET_MS" - The maximum time to wait for a DB connection in milliseconds. Defaults to `50`.
- "WATCHER_INFO_DB_POOL_QUEUE_INTERVAL_MS" - The interval in milliseconds to determine whether the queue target period above has been exceeded. Defaults to `1000`.

***Erlang VM configuration***

- "NODE_HOST" - The fully qualified host name of the current host.
- "ERLANG_COOKIE" - Magic cookie of the node.
- "REPLACE_OS_VARS" - An environment variable you export at runtime which instructed the tool to replace occurances of ${VAR} with the value from the system environment in the vm.args.

***Contract address configuration***
We allow a static configuration or a dynamic one, served as a http endpoint (one of them is mandatory).

- "ETHEREUM_NETWORK" - "RINKEBY" or "LOCALCHAIN".
- "CONTRACT_EXCHANGER_URL" - a server that can serve JSON in form of
```
{
  "plasma_framework_tx_hash":"<plasma_framework_tx_hash>",
  "plasma_framework":"<plasma_framework>",
  "eth_vault":"<eth_vault>",
  "erc20_vault":"<erc20_vault>",
  "payment_exit_game":"<payment_exit_game>",
  "authority_address":"<authority_address>"
}
```
Static configuration

- "ETHEREUM_NETWORK" - RINKEBY, ROPSTEN, MAINNET, or LOCALCHAIN
- "TXHASH_CONTRACT"
- "AUTHORITY_ADDRESS"
- "CONTRACT_ADDRESS_PLASMA_FRAMEWORK"
- "CONTRACT_ADDRESS_ETH_VAULT
- "CONTRACT_ADDRESS_ERC20_VAULT
- "CONTRACT_ADDRESS_PAYMENT_EXIT_GAME"

***Required contract addresses***

The contract addresses that are required to be included in the `contract_addr` field (or `_CONTRACT_ADDRESS` JSON) are:

```
{
  "plasma_framework": "...",
  "eth_vault": "...",
  "erc20_vault": "...",
  "payment_exit_game": "..."
}
```


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/omgnetwork/elixir-omg/tree/HEAD)

[Full Changelog](https://github.com/omgnetwork/elixir-omg/compare/v1.0.4...HEAD)

### Enhancements

- feat: handle in-flight exits deletions [\#1701](https://github.com/omgnetwork/elixir-omg/pull/1701) ([pgebal](https://github.com/pgebal))

### Bug fixes

- revert: reverts ife deletion commits [\#1725](https://github.com/omgnetwork/elixir-omg/pull/1725) ([pgebal](https://github.com/pgebal))
- fix: fix in-flight exit deleted bug and add tests [\#1714](https://github.com/omgnetwork/elixir-omg/pull/1714) ([pgebal](https://github.com/pgebal))
- fix: block submission stall monitor should ignore block\_submitting that are already mined [\#1703](https://github.com/omgnetwork/elixir-omg/pull/1703) ([unnawut](https://github.com/unnawut))

### Chores

- Allow to run docker-compose without feefeed docker [\#1726](https://github.com/omgnetwork/elixir-omg/pull/1726) ([boolafish](https://github.com/boolafish))
- feat: reintroduce automated changelog [\#1708](https://github.com/omgnetwork/elixir-omg/pull/1708) ([unnawut](https://github.com/unnawut))
- add feefeed docker to elixir-omg setup [\#1700](https://github.com/omgnetwork/elixir-omg/pull/1700) ([boolafish](https://github.com/boolafish))
- move omg\_performance json rpc tests to perf project [\#1691](https://github.com/omgnetwork/elixir-omg/pull/1691) ([ayrat555](https://github.com/ayrat555))

## [v1.0.4](https://github.com/omgnetwork/elixir-omg/tree/v1.0.4) (2020-09-03)

[Full Changelog](https://github.com/omgnetwork/elixir-omg/compare/v1.0.4-pre.1...v1.0.4)

### API changes

-  /block.validate endpoint [\#1668](https://github.com/omgnetwork/elixir-omg/pull/1668) ([okalouti](https://github.com/okalouti))

### Enhancements

- Block Validation: New Checks [\#1693](https://github.com/omgnetwork/elixir-omg/pull/1693) ([okalouti](https://github.com/okalouti))
- feat: configurable DB pool size, queue target and queue interval [\#1689](https://github.com/omgnetwork/elixir-omg/pull/1689) ([unnawut](https://github.com/unnawut))
- feat: block queue metrics and stalled submission alarm [\#1649](https://github.com/omgnetwork/elixir-omg/pull/1649) ([unnawut](https://github.com/unnawut))

### Bug fixes

- corrrectly serialize PIDs in alarms.get [\#1678](https://github.com/omgnetwork/elixir-omg/pull/1678) ([ayrat555](https://github.com/ayrat555))
- account.get\_exitable\_utxos is unaware of in-flight exited inputs [\#1676](https://github.com/omgnetwork/elixir-omg/pull/1676) ([pnowosie](https://github.com/pnowosie))
- Fix fee adapter to accept decimal value in fee rules [\#1662](https://github.com/omgnetwork/elixir-omg/pull/1662) ([jarindr](https://github.com/jarindr))
- Add missing clause on witness validation check [\#1656](https://github.com/omgnetwork/elixir-omg/pull/1656) ([mederic-p](https://github.com/mederic-p))
- fix: unexpected http method [\#1651](https://github.com/omgnetwork/elixir-omg/pull/1651) ([ripzery](https://github.com/ripzery))

### Chores

- bump version 1.0.4 [\#1722](https://github.com/omgnetwork/elixir-omg/pull/1722) ([jarindr](https://github.com/jarindr))
- auto trigger chart version bump [\#1695](https://github.com/omgnetwork/elixir-omg/pull/1695) ([boolafish](https://github.com/boolafish))
- bump phoenix [\#1680](https://github.com/omgnetwork/elixir-omg/pull/1680) ([InoMurko](https://github.com/InoMurko))
- chore: increase timeouts for childchain healthchecks [\#1671](https://github.com/omgnetwork/elixir-omg/pull/1671) ([ayrat555](https://github.com/ayrat555))
- fix integration tests [\#1654](https://github.com/omgnetwork/elixir-omg/pull/1654) ([ayrat555](https://github.com/ayrat555))
- feat: pin elixir and erlang versions for asdf [\#1648](https://github.com/omgnetwork/elixir-omg/pull/1648) ([unnawut](https://github.com/unnawut))
- chore: change log and version file change for v1.0.3 \(\#1638\) [\#1639](https://github.com/omgnetwork/elixir-omg/pull/1639) ([boolafish](https://github.com/boolafish))
- use cabbage tests from a separate repo [\#1636](https://github.com/omgnetwork/elixir-omg/pull/1636) ([ayrat555](https://github.com/ayrat555))
- set OMG.State GenServer timeout to 10s [\#1517](https://github.com/omgnetwork/elixir-omg/pull/1517) ([achiurizo](https://github.com/achiurizo))

### Documentation updates

- v.1.0.4 change log [\#1719](https://github.com/omgnetwork/elixir-omg/pull/1719) ([jarindr](https://github.com/jarindr))
- docs: extend description of running cabbage tests [\#1658](https://github.com/omgnetwork/elixir-omg/pull/1658) ([pnowosie](https://github.com/pnowosie))

## [v1.0.3](https://github.com/omgnetwork/elixir-omg/tree/v1.0.3) (2020-07-09)

[Full Changelog](https://github.com/omgnetwork/elixir-omg/compare/v1.0.3-pre.2...v1.0.3)

### API changes

- Add Transaction filter by end\_datetime [\#1595](https://github.com/omgnetwork/elixir-omg/pull/1595) ([jarindr](https://github.com/jarindr))

### Enhancements

- Add block processing queue to watcher info [\#1560](https://github.com/omgnetwork/elixir-omg/pull/1560) ([mederic-p](https://github.com/mederic-p))

### Bug fixes

- remove trace decorator from OMG.WatcherInfo.DB.EthEvent.get/1 [\#1640](https://github.com/omgnetwork/elixir-omg/pull/1640) ([ayrat555](https://github.com/ayrat555))
- get call\_data and rename it [\#1635](https://github.com/omgnetwork/elixir-omg/pull/1635) ([InoMurko](https://github.com/InoMurko))
- fix: handle "transaction underpriced" and other unknown server error responses [\#1617](https://github.com/omgnetwork/elixir-omg/pull/1617) ([unnawut](https://github.com/unnawut))

### Chores

- chore: change log and version file change for v1.0.3 [\#1638](https://github.com/omgnetwork/elixir-omg/pull/1638) ([boolafish](https://github.com/boolafish))
- sync v1.0.2 back to master [\#1626](https://github.com/omgnetwork/elixir-omg/pull/1626) ([boolafish](https://github.com/boolafish))

config/config.exs
import Config
ethereum_events_check_interval_ms = 8_000

config :logger, level: :info

config :logger, :console,
  format: "$date $time [$level] $metadata⋅$message⋅\n",
  discard_threshold: 2000,
  metadata: [:module, :function, :request_id, :trace_id, :span_id]

config :logger,
  backends: [Sentry.LoggerBackend, Ink]

config :logger, Ink,
  name: "elixir-omg",
  exclude_hostname: true,
  log_encoding_error: true

config :logger, Sentry.LoggerBackend,
  include_logger_metadata: true,
  ignore_plug: true

config :sentry,
  dsn: nil,
  environment_name: nil,
  included_environments: [],
  server_name: 'localhost',
  tags: %{
    application: nil,
    eth_network: nil,
    eth_node: :geth
  }

config :omg,
  deposit_finality_margin: 10,
  ethereum_events_check_interval_ms: ethereum_events_check_interval_ms,
  coordinator_eth_height_check_interval_ms: 6_000

config :omg, :eip_712_domain,
  name: "OMG Network",
  version: "1",
  salt: "0xfad5c7f626d80f9256ef01929f3beb96e058b8b4b0e3fe52d84f054c0e2a7a83"

config :omg_child_chain,
  submission_finality_margin: 20,
  block_queue_eth_height_check_interval_ms: 6_000,
  block_submit_every_nth: 1,
  block_submit_after_n_txs: :infinity,
  block_submit_max_gas_price: 20_000_000_000,
  metrics_collection_interval: 60_000,
  fee_adapter_check_interval_ms: 10_000,
  fee_buffer_duration_ms: 30_000,
  fee_adapter: {OMG.ChildChain.Fees.FileAdapter, opts: [specs_file_path: nil]}

config :omg_child_chain, OMG.ChildChain.Tracer,
  service: :omg_child_chain,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :omg_child_chain

# Configures the endpoint
# https://ninenines.eu/docs/en/cowboy/2.4/manual/cowboy_http/
# defaults are:
# protocol_options:[max_header_name_length: 64,
# max_header_value_length: 4096,
# max_headers: 100,
# max_request_line_length: 8096
# ]
config :omg_child_chain_rpc, OMG.ChildChainRPC.Web.Endpoint,
  render_errors: [view: OMG.ChildChainRPC.Web.Views.Error, accepts: ~w(json)],
  enable_cors: true,
  http: [:inet6, port: 9656, protocol_options: [max_request_line_length: 8192, max_header_value_length: 8192]],
  url: [host: "cc.example.com", port: 80],
  code_reloader: false

# Use Poison for JSON parsing in Phoenix
config :phoenix,
  json_library: Jason,
  serve_endpoints: true,
  persistent: true

config :omg_child_chain_rpc, OMG.ChildChainRPC.Tracer,
  service: :web,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :web

config :spandex_phoenix, tracer: OMG.ChildChainRPC.Tracer

config :omg_db,
  metrics_collection_interval: 60_000

ethereum_client_timeout_ms = 20_000

config :ethereumex,
  url: "http://localhost:8545",
  http_options: [recv_timeout: ethereum_client_timeout_ms]

config :omg_eth,
  contract_addr: nil,
  authority_address: nil,
  txhash_contract: nil,
  eth_node: :geth,
  child_block_interval: 1000,
  min_exit_period_seconds: nil,
  ethereum_block_time_seconds: 15,
  ethereum_events_check_interval_ms: ethereum_events_check_interval_ms,
  ethereum_stalled_sync_threshold_ms: 20_000,
  node_logging_in_debug: false

config :omg_status,
  statsd_reconnect_backoff_ms: 10_000,
  system_memory_check_interval_ms: 10_000,
  system_memory_high_threshold: 0.8

config :omg_status, OMG.Status.Metric.Tracer,
  service: :omg_status,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :backend

config :spandex, :decorators, tracer: OMG.Status.Metric.Tracer

config :statix,
  host: "datadog",
  port: 8125

config :spandex_datadog,
  host: "datadog",
  port: 8126,
  batch_size: 10,
  sync_threshold: 100,
  http: HTTPoison

config :vmstats,
  sink: OMG.Status.Metric.VmstatsSink,
  interval: 15_000,
  base_key: 'vmstats',
  key_separator: '$.',
  sched_time: true,
  memory_metrics: [
    total: :total,
    processes_used: :procs_used,
    atom_used: :atom_used,
    binary: :binary,
    ets: :ets
  ]

# Disable :os_mon's system_memory_high_watermark in favor of our own OMG.Status.Monitor.SystemMemory
# See http://erlang.org/pipermail/erlang-questions/2006-September/023144.html
config :os_mon,
  system_memory_high_watermark: 1.00,
  process_memory_high_watermark: 1.00

config :omg_watcher, child_chain_url: "http://localhost:9656"

config :omg_watcher,
  # 23 hours worth of blocks - this is how long the child chain server has to block spends from exiting utxos
  exit_processor_sla_margin: 23 * 60 * 4,
  # this means we don't want the `sla_margin` above be larger than the `min_exit_period`
  exit_processor_sla_margin_forced: false,
  maximum_block_withholding_time_ms: 15 * 60 * 60 * 1000,
  maximum_number_of_unapplied_blocks: 50,
  exit_finality_margin: 12,
  block_getter_reorg_margin: 200,
  metrics_collection_interval: 60_000

config :omg_watcher, OMG.Watcher.Tracer,
  service: :omg_watcher,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :omg_watcher

config :omg_watcher_info,
  child_chain_url: "http://localhost:9656",
  namespace: OMG.WatcherInfo,
  ecto_repos: [OMG.WatcherInfo.DB.Repo],
  metrics_collection_interval: 60_000,
  pending_block_processing_interval: 1000,
  block_queue_check_interval: 10_000

# Configures the endpoint

config :omg_watcher_info, OMG.WatcherInfo.DB.Repo,
  adapter: Ecto.Adapters.Postgres,
  # NOTE: not sure if appropriate, but this allows reasonable blocks to be written to unoptimized Postgres setup
  timeout: 180_000,
  connect_timeout: 180_000,
  url: "postgres://omisego_dev:omisego_dev@localhost/omisego_dev",
  migration_timestamps: [type: :timestamptz],
  telemetry_prefix: [:omg, :watcher_info, :db, :repo]

config :omg_watcher_info, OMG.WatcherInfo.Tracer,
  service: :ecto,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :db

# In mix environment, all modules are loaded, therefore it behaves like a watcher_info
config :omg_watcher_rpc,
  api_mode: :watcher_info

# Configures the endpoint
# https://ninenines.eu/docs/en/cowboy/2.4/manual/cowboy_http/
# defaults are:
# protocol_options:[max_header_name_length: 64,
# max_header_value_length: 4096,
# max_headers: 100,
# max_request_line_length: 8096
# ]
config :omg_watcher_rpc, OMG.WatcherRPC.Web.Endpoint,
  render_errors: [view: OMG.WatcherRPC.Web.Views.Error, accepts: ~w(json)],
  enable_cors: true,
  http: [:inet6, port: 7434, protocol_options: [max_request_line_length: 8192, max_header_value_length: 8192]],
  url: [host: "w.example.com", port: 80],
  code_reloader: false

config :phoenix,
  json_library: Jason,
  serve_endpoints: true,
  persistent: true

config :spandex_ecto, SpandexEcto.EctoLogger, tracer: OMG.WatcherInfo.Tracer

config :omg_watcher_rpc, OMG.WatcherRPC.Tracer,
  service: :web,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :web

config :spandex_phoenix, tracer: OMG.WatcherRPC.Tracer

config :briefly, directory: ["/tmp/omisego"]

# `watcher_url` must match respective `:omg_watcher_rpc, OMG.WatcherRPC.Web.Endpoint`
config :omg_performance, watcher_url: "localhost:7434"

import_config "#{Mix.env()}.exs"
